Tests/NIOExtrasTests/PCAPRingBufferTest.swift

```diff
@@ -0,0 +1,224 @@
+                XCTAssertEqual(tcpPayloadBytes, expectedData.readableBytes)
+            }
+        }())
+    }
+
+    class TriggerOnCumulativeSizeHandler : ChannelOutboundHandler {
+        typealias OutboundIn = ByteBuffer
+
+        private var bytesUntilTrigger: Int
+        private var pcapRingBuffer: NIOPCAPRingBuffer
+        private var sink: (ByteBuffer) -> ()
+
+        init(triggerBytes: Int, pcapRingBuffer: NIOPCAPRingBuffer, sink: @escaping (ByteBuffer) -> ()) {
+            self.bytesUntilTrigger = triggerBytes
+            self.pcapRingBuffer = pcapRingBuffer
+            self.sink = sink
+        }
+
+        func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+            if bytesUntilTrigger > 0 {
+                self.bytesUntilTrigger -= self.unwrapOutboundIn(data).readableBytes
+                if self.bytesUntilTrigger <= 0 {
+                    context.write(data).map {
+                        self.sink(captureBytes(ringBuffer: self.pcapRingBuffer))
+                    }.cascade(to: promise)
+                    return
+                }
+            }
+            context.write(data, promise: promise)
+        }
+    }
+
+    func testHandler() {
+        let maximumFragments = 3
+        let triggerEndIndex = 5
+        var testTriggered = false
+
+        func testRecordedBytes(buffer: ByteBuffer) {
+            var capturedData = buffer
+            XCTAssertNoThrow(try {
+                testTriggered = true
+                // See what we've got.
+                let data = dataForTests()
+                for expectedData in data[(triggerEndIndex - maximumFragments)..<triggerEndIndex] {
+                    var packet = capturedData.readPCAPRecord()
+                    let tcpPayloadBytes = try packet?.payload.readTCPIPv4()?.tcpPayload.readableBytes
+                    XCTAssertEqual(tcpPayloadBytes, expectedData.readableBytes)
+                }
+            }())
+        }
+
+
+        let channel = EmbeddedChannel()
+        let trigger = self.dataForTests()[0..<triggerEndIndex]
+            .compactMap { t in t.readableBytes }.reduce(0, +)
+
+        let pcapRingBuffer = NIOPCAPRingBuffer(maximumFragments: .init(maximumFragments),
+                                               maximumBytes: 1_000_000)
+        XCTAssertNoThrow(try channel.pipeline.addHandler(
+                            NIOWritePCAPHandler(mode: .client,
+                                                fileSink: pcapRingBuffer.addFragment),
+                name: "capture").flatMap {
+                    return channel.pipeline.addHandler(
+                        TriggerOnCumulativeSizeHandler(triggerBytes: trigger,
+                                                       pcapRingBuffer: pcapRingBuffer,
+                                                       sink: testRecordedBytes),
+                        name: "trigger")
+                }.wait())
```

Comment:
The formatting here is pretty crazy.

Splitting this into adding the handlers and then waiting on that future would help clear this up a bit.

Reply:
I have split - still looks pretty crazy but better than it was.